File: pkg/plugin/serve.go
// SPDX-License-Identifier: MIT

package plugin

import (
	"errors"
	"fmt"
	"io"
	"os"

	"github.com/adamtait/reviewer/pkg/finding"
)

// Handler is what a Go plugin implements.
//
// Describe is called once, after the handshake. Analyze is called once per
// analyzer the host chooses to run, never concurrently, and only for IDs that
// Describe reported. A handler may hold expensive state between Analyze calls —
// that is the reason the process is long-lived.
type Handler interface {
	// Name and Version identify the plugin in the run log and in version-skew
	// warnings.
	Name() string
	Version() string

	// Describe lists the analyzers this plugin provides in this repository. It is
	// the right place to decide Available: probing for an ESLint config here costs
	// one filesystem call and saves the host a pointless Analyze round trip.
	Describe() []Descriptor

	// Analyze runs one analyzer. Returning an error is not fatal to the run: the
	// host records it and continues without this analyzer's findings. Warnings are
	// for things worth saying that are not failures.
	Analyze(id string, req AnalyzeRequest) (findings []finding.Finding, warnings []string, err error)
}

// Serve runs the protocol conversation against r and w until the host says bye or
// the stream closes. It is the whole SDK: a Go plugin is a Handler plus
//
//	func main() { plugin.Serve(myHandler{}) }
//
// Serve writes protocol frames to w and nothing else, so a plugin must keep its
// own diagnostics off that stream.
func Serve(h Handler) error {
	return serve(h, os.Stdin, os.Stdout, os.Stderr)
}

func serve(h Handler, r io.Reader, w, logw io.Writer) error {
	in, out := NewReader(r), NewWriter(w)
	greeted := false

	for {
		f, err := in.Read()
		switch {
		case errors.Is(err, ErrClosed):
			// The host went away without saying bye. Not an error: it may have
			// timed out or been killed, and there is nothing to clean up.
			return nil
		case err != nil:
			// A line we cannot parse. Say so and keep reading: the host may
			// recover, and exiting would lose the analyzers we could still run.
			fmt.Fprintf(logw, "%s: %v\n", h.Name(), err)
			if writeErr := out.Write(Frame{Type: TypeError, Message: err.Error()}); writeErr != nil {
				return writeErr
			}
			continue
		}

		switch f.Type {
		case TypeHello:
			if f.Protocol != Protocol {
				// Refuse explicitly rather than guessing at an unknown dialect.
				err := fmt.Errorf("host speaks protocol %d, this plugin speaks %d", f.Protocol, Protocol)
				_ = out.Write(Frame{Type: TypeError, Message: err.Error()})
				return err
			}
			greeted = true
			if err := out.Write(Frame{
				Type:     TypeHello,
				Protocol: Protocol,
				Plugin:   h.Name(),
				Version:  h.Version(),
			}); err != nil {
				return err
			}

		case TypeDescribe:
			if !greeted {
				return errUngreeted(out, f.Type)
			}
			if err := out.Write(Frame{Type: TypeDescribe, Analyzers: h.Describe()}); err != nil {
				return err
			}

		case TypeAnalyze:
			if !greeted {
				return errUngreeted(out, f.Type)
			}
			req := AnalyzeRequest{}
			if f.Request != nil {
				req = *f.Request
			}
			findings, warnings, err := h.Analyze(f.Analyzer, req)
			if err != nil {
				fmt.Fprintf(logw, "%s: %s: %v\n", h.Name(), f.Analyzer, err)
				if err := out.Write(Frame{
					Type:     TypeError,
					Analyzer: f.Analyzer,
					Message:  err.Error(),
				}); err != nil {
					return err
				}
				continue
			}
			if err := out.Write(Frame{
				Type:     TypeFindings,
				Analyzer: f.Analyzer,
				Findings: findings,
				Warnings: warnings,
			}); err != nil {
				return err
			}

		case TypeBye:
			return nil

		default:
			// An unknown frame type from a newer host. Ignoring it is what keeps
			// the protocol additively extensible.
			fmt.Fprintf(logw, "%s: ignoring unknown frame type %q\n", h.Name(), f.Type)
		}
	}
}

func errUngreeted(out *Writer, got FrameType) error {
	err := fmt.Errorf("received %s before hello", got)
	_ = out.Write(Frame{Type: TypeError, Message: err.Error()})
	return err
}

File: plugins/typescript/src/serve.ts
// SPDX-License-Identifier: MIT

/**
 * The protocol loop.
 *
 * Channel discipline is the one rule that matters and the easiest to break:
 * **stdout carries frames and nothing else**. Anything a human should read goes
 * to stderr, which the host captures into the run log. A stray `console.log`
 * anywhere in this package — or in a library it calls — corrupts the stream, so
 * `console.log` is reassigned to stderr on startup rather than trusted not to be
 * called. ESLint and dependency-cruiser both print to stdout in some
 * configurations, which is not hypothetical.
 */

import { createInterface } from "node:readline";
import type { Readable, Writable } from "node:stream";
import { PROTOCOL, parseFrame, type AnalyzeRequest, type Descriptor, type Finding, type Frame } from "./protocol.js";

/** What an analyzer implementation provides. */
export interface Analyzer {
  readonly id: string;
  readonly order: number;
  readonly lane: "deterministic" | "llm";
  /**
   * Why this analyzer cannot run here, or undefined when it can. Called once,
   * during `describe`, so a repository missing an ESLint config is told at the
   * top of the run rather than per invocation.
   */
  unavailable?(root: string): Promise<string | undefined>;
  run(req: AnalyzeRequest): Promise<{ findings: Finding[]; warnings?: string[] }>;
}

export interface ServeOptions {
  name: string;
  version: string;
  analyzers: Analyzer[];
  input?: Readable;
  output?: Writable;
  log?: Writable;
}

/** Runs the conversation until the host says bye or the stream closes. */
export async function serve(opts: ServeOptions): Promise<void> {
  const input = opts.input ?? process.stdin;
  const output = opts.output ?? process.stdout;
  const log = opts.log ?? process.stderr;

  const write = (frame: Frame): void => {
    output.write(JSON.stringify(frame) + "\n");
  };
  const note = (message: string): void => {
    log.write(`${opts.name}: ${message}\n`);
  };

  const byId = new Map(opts.analyzers.map((a) => [a.id, a]));
  let greeted = false;

  const lines = createInterface({ input, crlfDelay: Infinity });
  for await (const line of lines) {
    if (line.trim() === "") continue;

    const parsed = parseFrame(line);
    if ("error" in parsed) {
      // Keep reading: the host may recover, and exiting would lose every
      // analyzer that could still have run.
      note(parsed.error);
      write({ type: "error", message: parsed.error });
      continue;
    }
    const frame = parsed.frame;

    switch (frame.type) {
      case "hello": {
        if (frame.protocol !== PROTOCOL) {
          const message = `host speaks protocol ${frame.protocol}, this plugin speaks ${PROTOCOL}`;
          write({ type: "error", message });
          note(message);
          return;
        }
        greeted = true;
        write({ type: "hello", protocol: PROTOCOL, plugin: opts.name, version: opts.version });
        break;
      }

      case "describe": {
        if (!greeted) return refuse(write, note, frame.type);
        write({ type: "describe", analyzers: await describe(opts.analyzers) });
        break;
      }

      case "analyze": {
        if (!greeted) return refuse(write, note, frame.type);
        const analyzer = frame.analyzer ? byId.get(frame.analyzer) : undefined;
        if (!analyzer) {
          write({
            type: "error",
            ...(frame.analyzer === undefined ? {} : { analyzer: frame.analyzer }),
            message: `no analyzer named ${String(frame.analyzer)}`,
          });
          break;
        }
        try {
          const { findings, warnings } = await analyzer.run(frame.request ?? { root: process.cwd(), changed: [] });
          write({ type: "findings", analyzer: analyzer.id, findings, ...(warnings?.length ? { warnings } : {}) });
        } catch (err) {
          // One analyzer failing is not fatal to the run (ADR-0013).
          const message = err instanceof Error ? err.message : String(err);
          note(`${analyzer.id}: ${message}`);
          write({ type: "error", analyzer: analyzer.id, message });
        }
        break;
      }

      case "bye":
        return;

      default:
        // A newer host may send frames this plugin predates. Ignoring them is
        // what keeps the protocol additively extensible.
        note(`ignoring unknown frame type ${JSON.stringify(frame.type)}`);
    }
  }
}

function refuse(write: (f: Frame) => void, note: (m: string) => void, got: string): void {
  const message = `received ${got} before hello`;
  write({ type: "error", message });
  note(message);
}

async function describe(analyzers: Analyzer[]): Promise<Descriptor[]> {
  const out: Descriptor[] = [];
  for (const a of analyzers) {
    const reason = await a.unavailable?.(process.cwd());
    out.push({
      id: a.id,
      lane: a.lane,
      order: a.order,
      available: reason === undefined,
      ...(reason === undefined ? {} : { unavailable: reason }),
    });
  }
  return out;
}

/**
 * Redirects console output to stderr. Called before any analyzer runs, because a
 * library that logs to stdout would otherwise be indistinguishable from a
 * protocol frame.
 */
export function protectStdout(): void {
  const toStderr = (...args: unknown[]): void => {
    process.stderr.write(args.map((a) => (typeof a === "string" ? a : JSON.stringify(a))).join(" ") + "\n");
  };
  console.log = toStderr;
  console.info = toStderr;
  console.debug = toStderr;
  console.warn = toStderr;
}
